### PR TITLE
[DataGrid] Remove `indeterminateCheckboxAction` prop

### LINKED ISSUE #14300: [data grid] Remove `indeterminateCheckboxAction` prop
Remove the prop `indeterminateCheckboxAction` and make `select` the default behavior.

Context: https://github.com/mui/mui-x/pull/14247#discussion_r1728468668

### PR DESCRIPTION
**Pull request by MBilalShafi**
Closes #14300

Todo:
- [x] Add to migration guide

Follow-ups:
- [ ] Codemod
- [ ] Cherry-pick `aria-label` fix to v7
- [ ] https://github.com/mui/mui-x/issues/15614

### Breaking changes

- The prop `indeterminateCheckboxAction` has been removed. Clicking on an indeterminate checkbox "selects" the unselected descendants.

### CODE REVIEW

File: packages/x-data-grid/src/models/props/DataGridProps.ts [hunk 1]
```diff
@@ -251,14 +251,6 @@ export interface DataGridPropsWithDefaultValues<R extends GridValidRowModel = an
    * @default false
    */
   ignoreDiacritics: boolean;
-  /**
-   * If `select`, a group header checkbox in indeterminate state (like "Select All" checkbox)
-   * will select all the rows under it.
-   * If `deselect`, it will deselect all the rows under it.
-   * Works only if `checkboxSelection` is enabled.
-   * @default "deselect"
```
Comment (flaviendelangle) [inline packages/x-data-grid/src/models/props/DataGridProps.ts:259]: Do we plan to deprecate on v7 next patch?

Comment (MBilalShafi) [reply packages/x-data-grid/src/models/props/DataGridProps.ts:259]: Seems a good idea to me. I'd be in favor of doing so, to let the v7 users mentally prepare for the breaking change beforehand.

Comment (MBilalShafi) [reply packages/x-data-grid/src/models/props/DataGridProps.ts:259]: Created https://github.com/mui/mui-x/issues/15614 to track it

**Review by arminmeh — APPROVED** [review]

**Comment by oliviertassinari** [comment]
Nice 👍

**Pull request merged**